### DISCUSSION TITLE
PorousFlow residual format

### DISCUSSION THREAD

**Discussion by Bearunner** [idaholab/moose#26334] [Q&A Modules: Porous Flow] [answered]
1

**Comment by cpgr** [comment, marked as answer]
The multiphase Darcy kernel (and all kernels derived from that) implement node-based upwinding, see https://mooseframework.inl.gov/modules/porous_flow/stabilization.html for lots of information. So there is lots of `nodenum` or `_i` instead of `_qp` in these kernels.

**Reply by Bearunner** [reply]
So there are two ways to write residuals: one is based on  [_qp] (e.g., [PorousFlowHeatConduction.C]), and another one is node number upwinding (e.g.,  PorousFlowHeatAdvection.C ) without using [_qp]?

What I know is that the integration of the quadrant points is very important in the finite element method. 

What you mean here is that the [_qp] in the residual can be replaced by the node number based on upwinding, right?

**Reply by cpgr** [reply]
If you look through that documentation, you see we lump mass to the nodes in the time derivative kernels, and therefore use the nodal fluxes in the advection kernels. There are benefits and also drawbacks as discussed - the important point is that we are consistent in how we do this. 

 Diffusion kernels and the like don't need this, so they are just formulated like normal using the quadrature points.

**Reply by Bearunner** [reply]
Thanks for your explanation.

1. According to what you said, my understanding is that one variable or property can have two formats in kernels depending on how we want to use them: one is node value and another is qp value， am I right?

2.  On line 94 in [PorousFlowDarcyBase.C](https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/kernels/PorousFlowDarcyBase.C#L94C1-L94C52)
  Real
  PorousFlowDarcyBase::darcyQp(unsigned int ph) const
  {
    return _grad_test[_i][_qp] *
         (_permeability[_qp] * (_grad_p[_qp][ph] - _fluid_density_qp[_qp][ph] * _gravity));
  }

  Why do we use qp value to calculate darcyQp, not node value as the mobility in this kernel? Can I also just 
   use node value for Darcy velocity calculation?


  Real
  PorousFlowDarcyBase::mobility(unsigned nodenum, unsigned phase) const
  {
    return _fluid_density_node[nodenum][phase] / _fluid_viscosity[nodenum][phase];
  }

**Reply by cpgr** [reply]
If I remember correctly, it is because the gradient of a variable (pressure) only exists at the qps, not the nodes
